### DISCUSSION TITLE
What are you doing with PSFalcon?

### DISCUSSION THREAD

**Discussion by bk-cs** [CrowdStrike/psfalcon#72] [Show and tell]
I'd love to hear your feedback on how you're using PSFalcon and what sort of additional commands or features you'd like to see. The module itself seems to be at a point where I can spend more time adding in additional functionality.

Two things I'm currently considering:

1. An `Uninstall-FalconSensor` command that's designed to remove the Falcon sensor (including getting the maintenance token) to specified host identifiers.
2. Incorporating support for the [Microsoft Secret Management](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.secretmanagement/?view=ps-modules) module for credential handling.

What are you using PSFalcon for today? What else would you like to see?

**Comment by z0her0** [comment]
Nice, both of those sound great.  

Sorry if I'm in the wrong forum, so feel free to relocate this if need:

I have a question on the demo you did with RiskIQ.  I installed PSRiskIQ, and I have the RiskIQ icon in CS, but is not integrated with Illuminate...  I'm getting errors when GetArtifacts.ps1 tries importing the custom IOC's into Falcon:
```
RiskIQ Article d8fef278 artifacts:

  9 domain result(s)
  40 hash_sha256 result(s)

Importing into CrowdStrike Falcon...
```
The first/original error (does not generate a csv):
```
.\GetArtifacts.ps1 -CSV .\passivetotal_article_d8fef278.csv

RiskIQ Article d8fef278 artifacts:

  9 domain result(s)
  40 hash_sha256 result(s)

Importing into CrowdStrike Falcon...New-FalconIoc : Cannot validate argument on parameter 'Array'. '2021-09-05T04:37:34.000Z' is not a valid 'expiration' value. 
{"action":"detect","description":"RiskIQ Article d8fef278","expiration":"2021-09-05T04:37:34.000Z","value":"53a48d3314d4c4660931653821496
415ba9687e4f14b458d9cf61cc96632c131","type":"sha256","applied_globally":true,"platforms":["windows","mac","linux"]}
At C:\Users\<redacted>\<redacted>\Documents\PSFalcon\GetArtifacts.ps1:146 char:38
+ ...               New-FalconIOC -Array $Array[$i..($i + ($MaxCount - 1))]
+                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [New-FalconIoc], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,New-FalconIoc
```
I removed the `expiration` field (this generated a csv, but still no upload to CS):
```
.\GetArtifacts.ps1 -CSV .\passivetotal_article_d8fef278.csv -Verbose

RiskIQ Article d8fef278 artifacts:

  9 domain result(s)
  40 hash_sha256 result(s)

Importing into CrowdStrike Falcon...VERBOSE: [ApiClient.Invoke] POST https://api.crowdstrike.com/oauth2/token
VERBOSE: [ApiClient.Invoke] ContentType: application/x-www-form-urlencoded, Accept: application/json
VERBOSE: [ApiClient.Invoke] 201: Created
VERBOSE: [ApiClient.Invoke] POST https://api.crowdstrike.com/iocs/entities/indicators/v1
VERBOSE: [ApiClient.Invoke] ContentType: application/json, Accept: application/json
VERBOSE: [ApiClient.Invoke] 400: BadRequest
Write-Result : 400: Provided data does not match expected format
At C:\Users\<redacted>\<redacted>\Documents\WindowsPowerShell\Modules\PSFalcon\2.1.2\Private\Private.ps1:582 char:35
+                         $Result = Write-Result $Request
+                                   ~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (System.Threadin...esponseMessage]:Task`1) [Write-Result], Exception
    + FullyQualifiedErrorId : be0a315e-e480-487d-9e8b-894a3c4c700a,Write-Result
```
I get the access token using my CS client ID/secret before running the script.  Do I need an API key from RiskIQ?  Or an enterprise account (not community)?  I recorded the second error with start/stop-transcript if you want that.

Any help is appreciated!  Thank you!

**Reply by bk-cs** [reply]
The error you're getting is because the CrowdStrike IOC format (and API) has changed since I created that script. I'll have to re-write it. I can do that next week.

**Reply by z0her0** [reply]
Nice!  Thank you much!

**Reply by bk-cs** [reply]
I think I responded having only read part of your error.

Yes, you do need the RiskIQ API key and need to be authenticated with PSRiskIQ before you're going to be able to pull the data from it, unless you download the indicators as CSV. 

It looks like the error is specifically tied to the format of the expiration time, with milliseconds being the problem. If you change line 137 in the script from:

```
                expiration       = $ExpirationDate.ToString('yyyy-MM-ddTHH:mm:ss.000Z')
```

to:

```
                expiration       = $ExpirationDate.ToString('yyyy-MM-ddTHH:mm:ssZ')
```

I think it'll fix the problem.

**Reply by z0her0** [reply]
Thank you, I will try that.   I tried changing the expiration format a couple different ways, then just removed it altogether but it didn't fix it.  I did not try `'yyyy-MM-ddTHH:mm:ssZ'` though.  Will let you know if it fixes it!

**Reply by z0her0** [reply]
That one gave me the same error.  I am going to run through the various RFC's and other time string format options (UnixDate, RubyDate, StampMilli, StampMicro, etc) until one works!

**Reply by bk-cs** [reply]
The format check is done by `New-FalconIoc` on import. It definitely matches that pattern, so maybe there's something that's not working in the conversion from CSV? Could you post an example CSV that will produce the error so I could try it out?

**Reply by z0her0** [reply]
This is the CSV I used (66 indicators total):
[passivetotal_article_d8fef278.csv](https://github.com/CrowdStrike/psfalcon/files/7125747/passivetotal_article_d8fef278.csv)

After removing the Expiration fields, this is the CSV output.  None of the URL indicator types converted (49 indicators remaining):
[2021-09-07_d8fef278.csv](https://github.com/CrowdStrike/psfalcon/files/7125766/2021-09-07_d8fef278.csv)

Maybe need to add 'url' to these areas?:
```
        # Retrieve artifacts from CSV
        Import-Csv $CSV | Where-Object { $_.type -match '(domain|hash_md5|hash_sha256|ip)' } |
            Select-Object type, value
```
and:
```
          # Convert type
            $Type = switch ($_.type) {
                'ip'          { 'ipv4' }
                'hash_md5'    { 'md5' }
                'hash_sha256' { 'sha256' }
                default       { $_ }
            }
```

This is the error I got when running the script without the Expiration:
```
.\GetArtifacts.ps1 -CSV .\passivetotal_article_d8fef278.csv -Verbose

RiskIQ Article d8fef278 artifacts:

  9 domain result(s)
  40 hash_sha256 result(s)

Importing into CrowdStrike Falcon...VERBOSE: [ApiClient.Invoke] POST https://api.crowdstrike.com/iocs/entities/indicators/v1
VERBOSE: [ApiClient.Invoke] ContentType: application/json, Accept: application/json
VERBOSE: [ApiClient.Invoke] 400: BadRequest
Write-Result : 400: One or more indicators have a warning or invalid input
At C:\Users\<redacted>\<redacted>\Documents\WindowsPowerShell\Modules\PSFalcon\2.1.2\Private\Private.ps1:582 char:35
+                         $Result = Write-Result $Request
+                                   ~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (System.Threadin...esponseMessage]:Task`1) [Write-Result], Exception
    + FullyQualifiedErrorId : ea3827bd-5c3b-4a4c-a292-1e7cddb06e92,Write-Result
 
complete.


    Directory: C:\Users\<redacted>\<redacted>\Documents\PSFalcon


Mode                 LastWriteTime         Length Name                                                                                   
----                 -------------         ------ ----                                                                                   
-a----          9/7/2021  11:45 PM           6469 2021-09-07_d8fef278.csv    
```

**Reply by bk-cs** [reply]
URLs are not accepted as custom IOCs -- you can do domains, IPv4 and IPv6 addresses, and SHA256 and MD5 hashes. 

Something about the indicators that are being submitted is not properly formatted, but it might be at a "warning" level rather than a "broken" level. You can change line 146 from:

```
New-FalconIOC -Array $Array[$i..($i + ($MaxCount - 1))]
```

to:

```
New-FalconIOC -Array $Array[$i..($i + ($MaxCount - 1))] -IgnoreWarnings $true
```

Which will allow the submission of the IOCs even if there is a warning.

**Reply by z0her0** [reply]
I tried your suggestion, and I also tried removing all the URLs from the passivetotal article before running GetArtifacts script.  Got same error both times, first line is:
```
Importing into CrowdStrike Falcon...Write-Result : 400: One or more indicators have a warning or invalid input
```
Looking at the generated CSV, [2021-09-08_d8fef278.csv](https://github.com/CrowdStrike/psfalcon/files/7132326/2021-09-08_d8fef278.csv), I think CrowdStrike is expecting a "severity" column to go along with "type", "value", "action", "platforms", "description", and "applied_globally".  I think if we can get "severity" column in there it may work.  Knock on wood.

**Reply by z0her0** [reply]
holy cow it worked @bk-cs

**Reply by z0her0** [reply]
Checked the IOC management UI and they all uploaded successfully.  Woot woot. @bk-cs 
```
RiskIQ Article d8fef278 artifacts:

  9 domain result(s)
  40 hash_sha256 result(s)

Importing into CrowdStrike Falcon...complete.
```

**Reply by bk-cs** [reply]
`-Severity` isn't required as far as I know, but something about creating those IOCs is still slightly off--though acceptable enough to create them if you ignore the warnings.

I'm glad it worked for you! If you figure out the exact field that was the problem, I'd be happy to fix it. I haven't been able to figure it out from glancing at the example CSVs.

**Reply by z0her0** [reply]
@bk-cs I emailed you

**Reply by z0her0** [reply]
Not sure if the email will get blocked because I attached the CSV of the riskIQ indicators... If you don't get it i'll resend with just the getartifacts.txt file @bk-cs

**Comment by z0her0** [comment]
Is there a way to download strict IOC's from a Falcon sandbox report > upload the IOC's into a threat indicator graph > then export the list of all hosts in contact with those IOC's?  I'm trying to figure out a smoother process of getting hostnames and host id's to use in PSFalcon without going into Host Management.  

Ideal workflow:  View Falcon sandbox report > click the 'Search Indicator Graph' button to see what hosts may be communicating with that malware sample > Refine indicator graph by removing noisey IOC's > Before exiting the "refine graph" interface, `'Export hosts to CSV'`.

Is this doable with PSFalcon?

**Reply by bk-cs** [reply]
You can't export a list from the Indicator Graph, or access the Indicator Graph using the API, but you can create the IOCs and set them to generate retroactive detections (which is effectively how they're displayed on the Indicator Graph).

You can get hostnames/host IDs using `Get-FalconHost`, and presumably do what you're trying to do using a combination of `Get-FalconReport`, and `New-FalconIoc` (with retroactive detections), then export the affected hosts based on the generated detections (`Get-FalconDetection`) or see which host IDs encountered the IOCs using `Get-FalconIocHost`.

**Reply by z0her0** [reply]
Oh awesome.  Good thinking.  I will give that a shot.
